File: Data Structures & Algorithms/validate-parentheses/submission-1.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        const stack = [];
        const map = {")": "(", "]": "[", "}": "{"};

        for (const c of s) {
            if (!Object.hasOwn(map, c)) {
                stack.push(c);
                continue;
            }
            if (stack.length === 0 || stack[stack.length - 1] != map[c]) return false;
            stack.pop();
        }

        return stack.length === 0;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [] # (i, h)
        res = 0

        for r, cur in enumerate(heights):
            start = r
            while stack and stack[-1][1] > cur:
                l, h = stack.pop()
                res = max(res, h * (r - l))
                start = l
            stack.append((start, cur))
        
        for i, h in stack:
            res = max(res, h * (len(heights) - i))
        
        return res
        
